feat: Enhanced Different TTLSecondsAfterFinished depending on if job is in Succeeded, Failed or Error, Fixes

NikeNano opened the pull request:
Checklist:

I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/1827) and discussed it with the community. 

Comment: This PR is note done but I have added it in order to get some more specific feedback and discuss the implementation.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/argoproj/argo?pullRequest=1883) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/argoproj/argo?pullRequest=1883) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/argoproj/argo?pullRequest=1883) it.</sub>

NikeNano commented:
Hi @sarabala1979 I have now made an initial PR with the workflow spec setup. I would be great if you could give some initial feedback. 

Thoughts: 
- I decided to leave TTLSecondsAfterFinished and added TTLStragety as : 
```
type TTLStrategy struct {
	SecondsAfterSuccess *int32 `json:"SecondsAfterSuccess:,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterSuccess:"`
	SecondsAfterFailed  *int32 `json:"SecondsAfterFailed,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterFailed"`
} 
```
My thought was that now there is now no change to the current design of TTLSecondsAfterFinished which should improve backward compatibility. 
- I also added so that priority is given for SecondsAfterSuccess and SecondsAfterFailed.  I updated with tests as well. 

Concerning the configmap for the controller. I have started to look for it, is it correct that what you mean by configmap for the controller is (this)[https://github.com/argoproj/argo/blob/ce78227abe5a3c901e5b7a7dd823fb2551dff584/workflow/config/config.go#L10]. 

If so how do you think is the base way to pass the information from the configmap object to the ttlcontroller? Atm as I understand the configmap is not passed to the ttlcontroller. 

Thanks for the feedback!

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
 
+// TTLStrategy is the strategy for the time to live depending on if the workflow succeded or failed
+type TTLStrategy struct {
+	SecondsAfterSuccess *int32 `json:"SecondsAfterSuccess:,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterSuccess:"`
```
Our convention has `json` variables be `camelCase`. Also, start with a fresh `protobuf` index for each `struct`
```suggestion
	SecondsAfterSuccess *int32 `json:"secondsAfterSuccess:,omitempty" protobuf:"bytes,1,opt,name=SecondsAfterSuccess:"`
```

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
+// TTLStrategy is the strategy for the time to live depending on if the workflow succeded or failed
+type TTLStrategy struct {
+	SecondsAfterSuccess *int32 `json:"SecondsAfterSuccess:,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterSuccess:"`
+	SecondsAfterFailed  *int32 `json:"SecondsAfterFailed,omitempty" protobuf:"bytes,18,opt,name=SecondsAfterFailed"`
```
Likewise
```suggestion
	SecondsAfterFailed  *int32 `json:"secondsAfterFailed,omitempty" protobuf:"bytes,2,opt,name=SecondsAfterFailed"`
```

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -201,6 +208,13 @@ type WorkflowSpec struct {
+	// deleted after the time to live expires. If this field is unset,
+	// the controller config map will hold the default values
+	// Update
+	TTLStrategy TTLStrategy `json:"TTLStrategy,omitempty" protobuf:"bytes,18,opt,name=TTLStrategy"`
```
```suggestion
	TTLStrategy TTLStrategy `json:"ttlStrategy,omitempty" protobuf:"bytes,28,opt,name=ttlStrategy"`
```

simster7 replied:
This field should also be of type `*TTLStrategy`

simster7 commented on workflow/ttlcontroller/ttlcontroller.go:
```diff
@@ -129,6 +129,9 @@ func (c *Controller) processNextWorkItem() bool {
 // enqueueWF conditionally queues a workflow to the ttl queue if it is within the deletion period
+// what is deletion period?
+// need to update here as well it seams like .....
+
```
Did you mean to delete these?

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -855,6 +869,11 @@ func (ws *WorkflowStatus) Successful() bool {
 	return ws.Phase == NodeSucceeded
 }
 
+// Failed return where or not the workflow has failed
```
```suggestion
// Failed return whether or not the workflow has failed
```

simster7 commented on workflow/ttlcontroller/ttlcontroller.go:
```diff
@@ -199,24 +202,43 @@ func (c *Controller) deleteWorkflow(key string) error {
+	}
 }
 
+// This is used to put thing on the que
```
Did you mean to delete this one too?

NikeNano commented:
Thanks for the feedback @simster7  I Will update based upon  your suggestions :)

NikeNano commented:
> Thanks for your PR @NikeNano! A couple of changes needed.
> 
> Also, I would suggest including `TTLStrategy.SecondsAfterCompleted` to replace `TTLSecondsAfterFinished`. This way, we can mark `TTLSecondsAfterFinished` as deprecated. To allow for backwards compatibility, we can automatically convert `TTLSecondsAfterFinished` to `TTLStrategy.SecondsAfterCompleted` during `Workflow` creation and provide a warning that it is being deprecated.
>

NikeNano commented:
FYI closed this by mistake.

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -201,6 +208,13 @@ type WorkflowSpec struct {
+	// deleted after the time to live expires. If this field is unset,
+	// the controller config map will hold the default values
+	// Update
+	TTLStrategy TTLStrategy `json:"ttlStrategy,omitempty" protobuf:"bytes,28,opt,name=ttlStrategy"`
```
> This should be of a `*TTLStrategy` type

simster7 commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -106,6 +107,12 @@ type WorkflowList struct {
 var _ TemplateStorage = &Workflow{}
 
+// TTLStrategy is the strategy for the time to live depending on if the workflow succeded or failed
+type TTLStrategy struct {
```
> Also, I would suggest including TTLStrategy.SecondsAfterCompleted to replace TTLSecondsAfterFinished. This way, we can mark TTLSecondsAfterFinished as deprecated. To allow for backwards compatibility, we can automatically convert TTLSecondsAfterFinished to TTLStrategy.SecondsAfterCompleted during Workflow creation and provide a warning that it is being deprecated.

NikeNano commented:
@simster7 I will fix it, havn't had time to look at it during Christmas:) Where in the code base is the workflow creation? I have looked around in the operation file but not really sure where the changes are needed in order to automatically convert TTLSecondsAfterFinished to TTLStrategy.SecondsAfterCompleted during Workflow creation? 

Hope you had a great Christamas!

NikeNano commented:
@simster7 I have now updated and the tests pass locally. However I have some issues understanding how to read the tests from the github action. The "default" gets me to a argo UI where I cant load the results.  The Build error gives "No configuration was found in your project. "

alexec commented:
You should be able to fix build we’re by syncing with master.

alexec commented:
The other error maybe a test failure. Run “make test” to check.

NikeNano commented:
make test

NikeNano commented:
> The other error maybe a test failure. Run “make test” to check.

Thanks! I assume you mean locally :) Running it and checking

NikeNano commented:
> The other error maybe a test failure. Run “make test” to check.

Running "make test" in the main folder did not return any errors. Is there any way to get the logs from the CI run?

NikeNano commented:
I have now updated according to master. Running make test locally the following test fails: argo/test/e2e it fails with No Auth Provider found for name "gcp". Is this something you are familiary with? I found this I https://github.com/kubernetes/client-go/issues/242 but don't know enough of the part of the repo to know if that makes sense .....? Is there any possibilities to get logs from the test run due to CI systems argo uses @alexec ?

NikeNano commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -201,6 +209,13 @@ type WorkflowSpec struct {
 	// ttlSecondsAfterFinished expires immediately after the Workflow finishes.
 	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty" protobuf:"bytes,18,opt,name=ttlSecondsAfterFinished"`
```
Sure I will fix this :) Was hoping to get the test running first but dont understand what the ```
hack/update-openapigen.sh
```
works, what is the goal of it? Currently get the following error 
```
API rule violation: names_match,github.com/argoproj/argo/pkg/apis/workflow/v1alpha1,TTLStrategy,Seco
```

simster7 replied:
You should make codegen using `make codegen`. You can ignore that API violation error

NikeNano commented:
> Last change, otherwise LGTM!

I updated with the tag

NikeNano commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -199,42 +207,50 @@ type WorkflowSpec struct {
 	// allowed to run before the controller terminates the workflow. A value of zero is used to
 	// terminate a Running workflow
-	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty" protobuf:"bytes,19,opt,name=activeDeadlineSeconds"`
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty" protobuf:"bytes,20,opt,name=activeDeadlineSeconds"`
```
Ahh ok, will fix that!

NikeNano replied:
I updated now, I might be wrong but I think is should be 28, based on that this is that last entry in 
```
type WorkflowSpec struct {
 ```
 https://github.com/argoproj/argo/blob/94449876f4a571ab279802d5ca4d5e938ca3d44d/pkg/apis/workflow/v1alpha1/workflow_types.go#L237  I am new to go and protobuff so not sure though :) Also thanks for all the feedback!

NikeNano commented:
> Thanks for all your work @NikeNano!

Awesome! Thanks for all the help as well @simster7!
